src/runtime/singlethreaded.rs

```diff
@@ -0,0 +1,83 @@
+use executor::current_thread::{self, CurrentThread};
+
+use tokio_reactor::{self, Reactor};
+use tokio_timer::timer::{self, Timer};
+use tokio_executor;
+
+use futures::Future;
+
+use std::io;
+
+/// Single-threaded runtime provides a way to start reactor
+/// and executor on the same thread
+///
+/// # Examples
+///
+/// Creating a new `Runtime` with default configuration values.
+///
+/// ```
+/// use tokio::runtime::SingleThreaded;
+/// use tokio::prelude::*;
+///
+/// let mut runtime = SingleThreaded::new().unwrap();
+///
+/// // Use the runtime...
+/// // runtime.block_on(f); // where f is a future
+/// ```
+#[derive(Debug)]
+pub struct SingleThreaded {
+    reactor_handle: tokio_reactor::Handle,
+    timer_handle: timer::Handle,
+    executor: CurrentThread<Timer<Reactor>>,
+}
+
+impl SingleThreaded {
+    /// Returns a new single-threaded runtime initialized with default
+    /// configuration values.
+    pub fn new() -> io::Result<SingleThreaded> {
+        // We need a reactor to receive events about IO objects from kernel
+        let reactor = Reactor::new()?;
+        let reactor_handle = reactor.handle();
+
+        // Place a timer wheel on top of the reactor. If there are no timeouts to fire, it'll let the
+        // reactor pick up new some new external events.
+        let timer = Timer::new(reactor);
+        let timer_handle = timer.handle();
+
+        // And now put a single-threaded executor on top of the timer. When there are no futures ready
+        // to do something, it'll let the timer or the reactor to generate some new stimuli for the
+        // futures to continue in their life.
+        let executor = CurrentThread::new_with_park(timer);
+
+        let runtime = SingleThreaded { reactor_handle, timer_handle, executor };
+        Ok(runtime)
+    }
+
+    /// Blocks and runs the given future.
+    ///
+    /// This is similar to running a runtime, but uses only the current thread.
+    pub fn block_on<F: Future<Item = (), Error = ()>>(&mut self, f: F) -> () {
+        let reactor_handle = self.reactor_handle.clone();
+        let timer_handle = self.timer_handle.clone();
+        // Binds an executor to this thread
+        let mut enter = tokio_executor::enter().expect("Multiple executors at once");
+
+        // This will set the default handle and timer to use inside the closure and run the future.
+        tokio_reactor::with_default(&reactor_handle, &mut enter, |enter| {
+            timer::with_default(&timer_handle, enter, |enter| {
+                // The TaskExecutor is a fake executor that looks into the current single-threaded
+                // executor when used. This is a trick, because we need two mutable references to the
+                // executor (one to run the provided future, another to install as the default one). We
+                // use the fake one here as the default one.
+                let mut default_executor = current_thread::TaskExecutor::current();
```

Comment:
Doesn't that mean we get two mutable references to it now? Wouldn't be a cleaner to e.g. get the executor from inside the `with_default` closure (where the other reference is not visible anymore), for example?

Reply:
Given there is no `unsafe` to see here this would be a bug in `current_thread`.  But only `Borrow::enter` actually sets the thread-local (which is used behind the scenes, `TaskExecutor` is just a stub struct), and it doesn't allow the mutable reference to escape while it's stored in there - looks ok to me.

What actually suprises me is that `Borrow::enter` doesn't also call `tokio_executor::with_default`: according to `TaskExecutor::current` docs it should only be called from within `run`, but that's not possible if we need to pass it to `tokio_executor::with_default`, so `current_thread` should set up that part itself (as `TaskExecutor` is a stub struct right now it works anyway).

And the bad news is: changing it is difficult, as `tokio_executor::with_default` doesn't like getting called nested... perhaps cleanup `current_thread` first (also see my comment about forwarding park errors; the error handling needs to be redone imho).